### ISSUE TITLE
Cannot debug my extension on this windows user profile anymore

### ISSUE THREAD

**Issue by HenrikHolmIT** [microsoft/vscode#84685]
Issue Type: <b>Bug</b>

I have no idea if this can be replicated. For some odd reason I cannot debug my extension. When I hit F5 VS Code keeps Building forever. Im never presented with an error message. Stop button at the top is visible. The spinning Building symbol is visible at the buttom but nothing happens. If I click the Stop button the debug bar goes away but the Building message keeps spinning until I restart VS Code. Ive reinstalled VS Code and downgraded to older versions. This still happens. Ive tried to hard reset my VS Code profile on this user by uninstalling and deleting all data storage locations. Still the same.

I then tried  to create a new user on my windows machine. I started VS Code and with that new user Im able to debug my extension so obviously its my profile somewhere that is messed up. It would be nice to be notified of what is wrong...

Im willing to provide as much info as humanly possible. I very much like to be able to debug my extension on this windows profile instead of having to log in as another user when I need to update something or add a new feature.

Running Windows 10 Pro 1809 with Visual Studio 2017 and Visual Studio 2019 Community Edition.

Version: 1.40.0 (user setup)
Commit: 86405ea23e3937316009fc27c9361deee66ffbf5
Date: 2019-11-06T17:02:13.381Z
Electron: 6.1.2
Chrome: 76.0.3809.146
Node.js: 12.4.0
V8: 7.6.303.31-electron.0
OS: Windows_NT x64 10.0.17763

VS Code version: Code 1.40.0 (86405ea23e3937316009fc27c9361deee66ffbf5, 2019-11-06T17:02:13.381Z)
OS version: Windows_NT x64 10.0.17763

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz (8 x 1992)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>protected_video_decode: enabled<br>rasterization: enabled<br>skia_renderer: disabled_off<br>surface_control: disabled_off<br>surface_synchronization: enabled_on<br>video_decode: enabled<br>viz_display_compositor: enabled_on<br>viz_hit_test_surface_layer: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|7.82GB (1.04GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (2)</summary>

Extension|Author (truncated)|Version
---|---|---
vscode-markdownlint|Dav|0.32.0
vscode-typescript-tslint-plugin|ms-|1.2.3

</details>

**Comment by ThomasGri** [comment]
Hello Henrik,

I have the exact same problem. it appeared yesterday. For me the problem is a "watch" task of npm. If I terminate the terminal after hitting f5 i can debug the extension.

Hope this can help as a workaround.

**Comment by HenrikHolmIT** [comment]
Works. What can be causing this?

I mean this line was also available in the project when I debug with my other windows user profile on which everything worked fine...

I tried your solution of terminating the terminal and the debugging starts up. I am notified that 'The terminal process terminated with exit code: 1' but at least now I can work on the extension from my main windows user profile.

Thanks

**Comment by HenrikHolmIT** [comment]
If I remove this line from launsh.json it runs as usual.

"preLaunchTask": "npm: watch"

**Comment by IllusionMH** [comment]
/needsMoreInfo

If you have `preLaunch` task VS Code will wait until it ends before starting debug, but it never ends for watch tasks so there won't be a signal to start debug unless you set up problem matchers.

Do you have problem matchers configured as shown discussed in https://github.com/microsoft/vscode/issues/6209?

If there are no problem matchers you'll need to add them or launch watcher separately.

**Comment by HenrikHolmIT** [comment]
@IllusionMH But why is it suddenly not existing?

I made an update on the extension November first and had zero problems. I've never edited this file...

I do have a preLLaunchTask: "preLaunchTask": "npm: watch"
And my watch task is: "watch": "tsc -watch -p ./"

This was also the setup November 1st. No change was made to this behavior but it worked.

I have no problem removing it if it serves no purpose.

**Comment by alexr00** [comment]
On `watch` prelaunch tasks we use a `problemMatcher` to parse the output from the task and determine when the task is "done" based on the regex specified in the `problemMatcher`. With an "npm: watch" task, we wouldn't know to automatically include a problem matcher, so you'll need to do that yourself in your task.

Is it possible that you didn't have a `prelaunchTask` before? Is it possible your `watch` script in your package.json changed?

Since your `prelaunchTask` is a tsc watch task, you'll likely need to have this in your tasks.json:
```
{
	"version": "2.0.0",
	"tasks": [
		{
			"type": "npm",
			"script": "watch",
			"problemMatcher": "$tsc-watch",
			"isBackground": true,
		}
	]
}
```

**Comment by HenrikHolmIT** [comment]
@alexr00 I can confirm that there is a problemMatcher in task.json

I can also confirm that all these parts have always been the way they are now ever since the extension was first created. I am the only developer to have worked on it. The only thing ever changed or added to package.json was a 'build' script to perform a 'vsce package' but these days I just use 'vsce package' when ever I need to build an vsix file for users of the extension.

`{
	"version": "2.0.0",
	"tasks": [
		{
			"type": "npm",
			"script": "watch",
			"problemMatcher": "$tsc-watch",
			"isBackground": true,
			"presentation": {
				"reveal": "never"
			},
			"group": {
				"kind": "build",
				"isDefault": true
			}
		}
	]
}`

**Comment by alexr00** [comment]
Since the problem matchers rely on regular expressions, if the output of your task changed the the problem matcher might not be able to find an "end" to your task. Can you share the output from running that task?

And, just for completeness, can you share your the launch.json configuration that's having this problem?

**Comment by alexr00** [comment]
Also, since you say this is user specific, can you share your user settings.json for VS Code?

**Comment by HenrikHolmIT** [comment]
The problem I think is that the watch does not produce any output. I dont know where to find the watch output other than Terminal tab Task - watch? If that is the correct place this is what I see:

> Executing task: npm run watch <

> smartcode@0.0.24 watch c:\Development\Repos\SMARTcode
> tsc -watch -p ./

I can say that I have to scroll way up high in the tab to see this text. As if many empty lines have been output to the page?

launch.json
```json
{
	"version": "0.2.0",
	"configurations": [
		{
			"name": "Run Extension",
			"type": "extensionHost",
			"request": "launch",
			"runtimeExecutable": "${execPath}",
			"args": [
				"--extensionDevelopmentPath=${workspaceFolder}"
			],
			"outFiles": [
				"${workspaceFolder}/out/**/*.js"
			],
			"preLaunchTask": "npm: watch"
		},
		{
			"name": "Extension Tests",
			"type": "extensionHost",
			"request": "launch",
			"runtimeExecutable": "${execPath}",
			"args": [
				"--extensionDevelopmentPath=${workspaceFolder}",
				"--extensionTestsPath=${workspaceFolder}/out/test"
			],
			"outFiles": [
				"${workspaceFolder}/out/test/**/*.js"
			],
			"preLaunchTask": "npm: watch"
		}
	]
}
```
C:\Users\username\AppData\Roaming\Code\User\settings.json
```json
{
    "git.autofetch": true,
    "git.enableSmartCommit": true,
    "terminal.integrated.shell.windows": "C:\\Program Files\\Git\\bin\\bash.exe",
    "git.confirmSync": false,
    "debug.onTaskErrors": "debugAnyway"
}
```

If I create a new extension with 'yo code' command it will produce the same result. It will hang when debugging so it may not be needed to share repository?

**Comment by alexr00** [comment]
> I can say that I have to scroll way up high in the tab to see this text. As if many empty lines have been output to the page?

This is key. It sounds like you're hitting a known terminal conpty issue, but I'm not sure how given your Windows build is 17763. @Tyriar could this be https://github.com/microsoft/vscode/issues/84082?

**Comment by HenrikHolmIT** [comment]
VS Code updated this morning and I still have the problem.

Version: 1.40.1 (user setup)
Commit: 8795a9889db74563ddd43eb0a897a2384129a619
Date: 2019-11-13T16:49:35.976Z
Electron: 6.1.2
Chrome: 76.0.3809.146
Node.js: 12.4.0
V8: 7.6.303.31-electron.0
OS: Windows_NT x64 10.0.17763

**Comment by Tyriar** [comment]
@alexr00 sounds like it, @HenrikHolmIT is on 17763 which conpty doesn't work on. Need Windows 1903+ (build 18309+ I think).

**Comment by alexr00** [comment]
@Tyriar the issue still happens on 1.40.1, where conpty should be correctly not used again. Am I understanding #84082 correctly?

**Comment by ejgallego** [comment]
> €: It seems to be that exactly this line inside my package.json is causing the problem:
> "watch": "tsc -watch -p ./",
> 
> After removing it the Debugger starts without any problem.

Bingo! I had the exact same problem after updating to 1.40 ; using Linux tho. Will modify my debug profile for now.

**Comment by qjebbs** [comment]
Same here with VSC1.40-1.40.1 on Mac 10.14.6, that's why I use `"preLaunchTask": "npm: compile"` to bypass the issue.

**Comment by alexr00** [comment]
@qjebbs are you seeing a lot of unexpected empty lines in your tasks terminal too?

**Comment by qjebbs** [comment]
@alexr00  Not exactly, all I can see is (in terminal, nothing displayed in task output panel):

```log
[下午4:44:28] Starting compilation in watch mode...

[下午4:44:31] Found 0 errors. Watching for file changes.

```

no many-of-empty-lines.

No matter in what way, run in terminal: `node_modules/typescript/bin/tsc  -watch -p ./`, or run the watch task with command palette

**Comment by qjebbs** [comment]
More to supplement about the problem I encountered:

1. With VSCode 1.40.1 newly started, press `F5` to start up the debug work, a `jebbs.plantuml` extension. Debug window shows.
2. Click to restart debug session in debug toolbar, the debug window won't show up again.
3. Try to close VSCode, no response. I have to force quit VSCode, or kill the watching process (`npm`) before I can quit.

**Comment by alexr00** [comment]
The problem is that the typescript problem matcher is not localizable. So: 
```
[下午4:44:28] Starting compilation in watch mode...

[下午4:44:31] Found 0 errors. Watching for file changes.
```
You have some characters before the timestamp there that the `$tsc-watch` problem matcher can't deal with. I'm sad to say I see two options that are probably not ideal for you:
- Change your system language to a language that the `$tsc-watch` problem matcher can handle, such as English or another that formats the time as HH:MM:SS AM/PM.
- Modify the problem matcher to suite your needs better. You can use the following as a starting point, but you'll need to change the `beginsPattern` and `endsPattern` regexp to match your timestamp format since this one uses the English matching timestamp format:
```
"problemMatcher": {
				"base": "$tsc-watch",
				"background": {
					"beginsPattern": {
						"regexp": "^\\s*(?:message TS6032:|\\[?\\d{1,2}:\\d{1,2}:\\d{1,2}(?: AM| PM| a\\.m\\.| p\\.m\\.)?(?:\\]| -)) File change detected\\. Starting incremental compilation\\.\\.\\."
					  },
					  "endsPattern": {
						"regexp": "^\\s*(?:message TS6042:|\\[?\\d{1,2}:\\d{1,2}:\\d{1,2}(?: AM| PM| a\\.m\\.| p\\.m\\.)?(?:\\]| -)) (?:Compilation complete\\.|Found \\d+ errors?\\.) Watching for file changes\\."
					  }				
				}
			},
```
Duplicate of https://github.com/microsoft/vscode/issues/46373

**Comment by HenrikHolmIT** [comment]
@alexr00 

I did not change language. It used to work on the language I am now experiencing the issue on. Then one day it stopped working. Was this case supposed to be closed? 

I still have the issue though :(

**Comment by Rehan-MALAK** [comment]
I was able to make @ejgallego plugin work (https://github.com/Deducteam/lambdapi/pull/240) by taking the .vscode/ directory builded with ``yo code`` from the tutorial https://code.visualstudio.com/api/get-started/your-first-extension

.vscode/extensions.json
.vscode/launch.json
.vscode/settings.json
.vscode/tasks.json

Hope that helps

**Comment by qjebbs** [comment]
@alexr00 Thanks for you detailed solution. But it's weird that I never come into this issue before I updated to `1.40.0`.

**Comment by HenrikHolmIT** [comment]
Working around this issue has become a nightmare.
My process:
I press F5 to debug
Wait a second and remember that I have to click Building text in the status bar
Then click the Kill Terminal trashcan
Wait for the Debug VS Code to open
Start to debug code
Do the actions I need to do to get to the code I want to debug
Realize the code is not compiled with the latest changes so I start over
Repeat...

Sometimes two VSCode Windows opens up on debug. Close them and try again. But then the Workspace I debug in is not loaded anymore and I have to browse for it. Then start over to realize I'm still not working on the latest code. Repeat again.

This is horribly effective for my productivity in that I get nothing done.

Currently upgrading to Windows 10 1909. Hope it get rid of this annoying process...

**Comment by alexr00** [comment]
I made the typescript problem matcher less local specific in https://github.com/microsoft/vscode/commit/e82dd213dde5a1b042d6c2e65dd7bd49345c683b
Can you try installing [VS Code insiders ](https://code.visualstudio.com/insiders/)and see if that works for you?

**Comment by qjebbs** [comment]
@alexr00 Thanks for the work, I'll try it. But I'm just wondering why don't to make it more flexible by (focus on the part isn't likely changing):

```json
"problemMatcher":  {
	"base": "$tsc-watch",
	"background": {
		"beginsPattern": {
			"regexp": "Starting compilation in watch mode|File change detected\\.\\s*Starting incremental compilation"
		},
		"endsPattern": {
			"regexp": "(?:Compilation complete\\.|Found \\d+ errors?\\.)\\s*Watching for file changes\\."
		}
	}
},
```

This works for me.

**Comment by qjebbs** [comment]
PS: With the current stable release, I change the `tasks.json` by what `RegExp` you committed:

```json
"background": {
	"beginsPattern": {
		"regexp": "^\\s*(?:message TS6032:|\\[?\\D*\\d{1,2}:\\d{1,2}:\\d{1,2}\\D*(?:\\]| -)) File change detected\\. Starting incremental compilation\\.\\.\\."
	},
	"endsPattern": {
		"regexp": "^\\s*(?:message TS6042:|\\[?\\D*\\d{1,2}:\\d{1,2}:\\d{1,2}\\D*(?:\\]| -)) (?:Compilation complete\\.|Found \\d+ errors?\\.) Watching for file changes\\."
	}
}
```

It doesn't work, and prompts "cannot tract task" or something (in Chinese). I think the insider release should have same result, doesn't it?

**Comment by HenrikHolmIT** [comment]
@alexr00 Updated to 1909. Seems the problem has god away. F5 opens a debug VSCode window now and it seems to always run the latest code.

I'm happy now.

**Comment by alexr00** [comment]
@qjebbs I think you'd need to add a bit more to the task:
```
 "problemMatcher": {
				"base": "$tsc-watch",
				"background": {
					"beginsPattern": {
                                                 "regexp": "^\\s*(?:message TS6032:|\\[?\\D*\\d{1,2}:\\d{1,2}:\\d{1,2}\\D*(?:\\]| -)) File change detected\\. Starting incremental compilation\\.\\.\\."
                                          },
                                          "endsPattern": {
                                                 "regexp": "^\\s*(?:message TS6042:|\\[?\\D*\\d{1,2}:\\d{1,2}:\\d{1,2}\\D*(?:\\]| -)) (?:Compilation complete\\.|Found \\d+ errors?\\.) Watching for file changes\\."
                                           }			
				}
			},
``` 
Insiders installs side by side with stable, so it's pretty easy to try out.

**Comment by qjebbs** [comment]
@alexr00 I think the insider has fixed this, according to my test. 👍

**Comment by alexr00** [comment]
Excellent, thank you for checking!
